compiler/src/iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp

```diff
@@ -735,32 +752,34 @@ struct FlattenMemRefSubspanPass
 
     RewritePatternSet flattenPatterns(context);
 
-    // For subspan ops that "generate" MemRef values, convert all MemRef types
-    // to 1-D dynamic sized one. This matches how IREE models buffers nicely.
-    FlattenMemRefTypeConverter fullyDynamicTypeConverter;
-    flattenPatterns.add<FlattenBindingSubspan>(fullyDynamicTypeConverter,
-                                               context);
-
-    // For other ops that generate MemRef values, we may not be able to go fully
-    // dynamic (e.g., memref::GlobalOp). Still convert everything to 1-D though.
-    FlattenMemRefTypeConverter only1DStaticTypeConverter;
-    only1DStaticTypeConverter.addConversion(
-        [](MemRefType type) -> Optional<Type> {
-          // 1-D MemRef types are okay.
-          if (isRankOneMemRef(type)) return type;
-
-          // Fall back to the default conversion flow.
-          return llvm::None;
-        });
+    // Interface binding subspan ops represents allocations from the runtime. We
+    // convert all MemRef types "generated" by them to 1-D one, static for
+    // uniform buffers and dynamic for storage buffers. This matches how IREE
+    // models runtime buffers nicely.
+    FlattenMemRefTypeConverter interfaceTypeConverter;
```

Comment:
This is not really realted to this change, but why are they two different type converters here?

Reply:
Actually the comments there are trying to explain that. More concretely, it's due to that we need to differentiate memref generated from hal interface subspan ops vs. other ops like memref.alloc for shared memory. For the former we always convert them into 1-D dynamic case for storage buffers (and this one enables converting to 1-D static case for uniform buffers), so that we can unify different subspan ops into the same underlying storage buffer (for SPIR-V at least), with the subspan op offset moved into load/store. While for the latter we need to preserve the static size for allocation. It's a bit unfortunate with this complication. Actually once we can plumb `#hal.descriptor_type` through memref memory space https://github.com/iree-org/iree/pull/11227, this can be cleaned up because we will have ways to differentiate memrefs generated from different ops via the memory space attribute. Then we can unify into one type converter.

Reply:
Hmmm, I am not sure I follow.... They both seem to be doing the same.... anyway, I dont know the details of this pass off the top of my head.

Reply:
Yeah, it's confusing I know. I'll take another look after landing #11227 to clean up this hopefully.